src/main/java/at/nonblocking/maven/nonsnapshot/impl/UpstreamDependencyHandlerDefaultImpl.java

```diff
@@ -48,54 +50,131 @@
+      for(MavenModuleDependency indirectDependency : bomMavenModule.getDependencies()) {
+        MavenArtifact indirectDependencyArtifact = indirectDependency.getArtifact();
+        String indirectDependencyScope = indirectDependency.getScope();
+        processedUpstreamDependencies.add(makeProcessedUpstreamDependencyFromMavenArtifact(indirectDependencyArtifact));
+        LOG.info(
+          "Indirect dependency: {} (scope: {}) found and added", indirectDependencyArtifact.toString(), indirectDependencyScope);
+        if(indirectDependencyArtifact.getType() != null && indirectDependencyScope != null &&
+           indirectDependencyArtifact.getType().equals("pom") && indirectDependencyScope.equals("import"))
+        {
+          LOG.info(
+            "Indirect dependency: {} is of pom type and has import scope - it will be treated as bom upstream dependency",
+            indirectDependencyArtifact.toString()
+          );
+          ProcessBomUpstreamDependency(
+            indirectDependencyArtifact.toString(), processedUpstreamDependencies, mavenPomHandler, repositorySystem, repositorySystemSession, remoteRepositories);
+        }
+
 
-    for (String upstreamDependencyString : upstreamDependencyStrings) {
-      if (upstreamDependencyString.trim().isEmpty()) {
-        throw new NonSnapshotPluginException("Illegal upstreamDependency: " + upstreamDependencyString);
       }
+    } catch (ArtifactResolutionException e) {
+      throw new NonSnapshotDependencyResolverException("Couldn't resolve bom dependency: " + upstreamDependencyString, e);
+    }
+  }
 
-      Pattern groupPattern;
-      Pattern artifactPattern = null;
-      Integer versionMajor = null;
-      Integer versionMinor = null;
-      Integer versionIncrement = null;
+  @Override
+  public List<ProcessedUpstreamDependency> processDependencyList(List upstreamDependencies,
+                                                                 MavenPomHandler mavenPomHandler,
+                                                                 RepositorySystem repositorySystem,
+                                                                 RepositorySystemSession repositorySystemSession,
+                                                                 List<RemoteRepository> remoteRepositories) {
+    LOG.info("Collecting upstream dependencies");
 
-      String[] parts = upstreamDependencyString.split(":");
-      groupPattern = createPattern(parts[0]);
-      if (parts.length > 1) {
-        artifactPattern = createPattern(parts[1]);
+    if (upstreamDependencies == null || upstreamDependencies.isEmpty()) {
+      return null;
+    }
+
+    List<ProcessedUpstreamDependency> processedUpstreamDependencies = new ArrayList<>(upstreamDependencies.size());
+
+    for (Object upstreamDependency : upstreamDependencies) {
+      boolean isUpstreamBomDependency = false;
+      String upstreamDependencyString;
+
+      if(upstreamDependency instanceof BomDependency) {
```

Comment:
The idea of storing strings and `BomDependency` objects in the same list is not intuitive in my opinion. It took me quite some time to understand what's going on.

In fact, `BomDependency` class is just a wrapper for String. The only difference is that has a different type and you can use `instanceof` to tell which is which.
I think making a generic `UpstreamDependency` class with a boolean (or better - enum) member to tell if it's a regular dependency or a BOM would be a more readable solution.
Another solution would be to provide to separate lists of upstream dependencies and bom dependencies to this function.

Reply:
bbad, please read first about this guide about rules related to adding parameters to maven plugins: https://maven.apache.org/guides/mini/guide-configuring-plugins.html There you can find information for what reason BomDependency class was created and where it is used.
Secondary, please take a look on README.md, where you can find:
```
The upstream dependency list is processed in order of their definition and the first match is taken. That allows it to define an exceptions from a wildcard rule like this:
```
and please understand that ```Another solution would be provide to separate lists of upstream dependencies and bom dependencies to this function.`` is not acceptable cause it will change that behavior of plugin. If we want to contribute with that changes to origin solution we shouldn't change this.